src/main/java/harmonised/pmmo/mixin/EnchantApplyMixin.java

```diff
@@ -1,33 +1,21 @@
 package harmonised.pmmo.mixin;
 
-import java.util.List;
-
-import org.spongepowered.asm.mixin.Debug;
+import net.minecraft.advancements.critereon.EnchantedItemTrigger;
+import net.minecraft.server.level.ServerPlayer;
 import org.spongepowered.asm.mixin.Mixin;
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.Inject;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
-import org.spongepowered.asm.mixin.injection.callback.LocalCapture;
 
 import harmonised.pmmo.api.events.EnchantEvent;
-import net.minecraft.core.BlockPos;
-import net.minecraft.world.entity.player.Player;
-import net.minecraft.world.inventory.EnchantmentMenu;
 import net.minecraft.world.item.ItemStack;
-import net.minecraft.world.item.enchantment.EnchantmentInstance;
-import net.minecraft.world.level.Level;
 import net.minecraftforge.common.MinecraftForge;
 
-@Debug
-@Mixin(EnchantmentMenu.class)
-public class EnchantApplyMixin {	
-
-	@Inject(method = "lambda$clickMenuButton$1(Lnet/minecraft/world/item/ItemStack;ILnet/minecraft/world/entity/player/Player;ILnet/minecraft/world/item/ItemStack;Lnet/minecraft/world/level/Level;Lnet/minecraft/core/BlockPos;)V",
```

Comment:
This mixin breaks a lot and that is a caveat I have to tell people about this event.  Your implementation is a more stable approach from the mixin perspective, but it fundamentally changes how the pmmo enchant behavior works.  

With my implementation the player will receive XP for every enchantment that is applied and earn XP based on that enchantment level.  With your implementation the player would earn the full XP specified on the item times the levels spent (which is typically only 1,2, or 3) regardless of what the enchantment is.  This would break existing user configurations and modpacks.

Lastly, the reason my mixin breaks is because other mods mix into the enchant menu and add locals.  I can't control that and there's no way to work around it other than to let users know when it comes up.   This is why i made a PR into NeoForge for 1.21 to make sure the enchant behavior was patched into the game so i wouldn't need the mixin going forward.

Reply:
I can give it another try in a couple days, this time without modifying the event :)